protocol/nats_jetstream/v3/message.go

```diff
@@ -0,0 +1,145 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/nats-io/nats.go/jetstream"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/binding/format"
+	"github.com/cloudevents/sdk-go/v2/binding/spec"
+)
+
+const (
+	// see https://github.com/cloudevents/spec/blob/main/cloudevents/bindings/nats-protocol-binding.md
+	prefix            = "ce-"
+	contentTypeHeader = "content-type"
+)
+
+var specs = spec.WithPrefix(prefix)
+
+// Message implements binding.Message by wrapping an jetstream.Msg.
+// This message *can* be read several times safely
+type Message struct {
+	Msg      jetstream.Msg
+	encoding binding.Encoding
+}
+
+// NewMessage wraps an *nats.Msg in a binding.Message.
+// The returned message *can* be read several times safely
+// The default encoding returned is EncodingStructured unless the NATS message contains a specversion header.
+func NewMessage(msg jetstream.Msg) *Message {
+	encoding := binding.EncodingStructured
+	if msg.Headers() != nil {
+		if msg.Headers().Get(specs.PrefixedSpecVersionName()) != "" {
+			encoding = binding.EncodingBinary
+		}
+	}
+	return &Message{Msg: msg, encoding: encoding}
+}
+
+var _ binding.Message = (*Message)(nil)
+
+// ReadEncoding return the type of the message Encoding.
+func (m *Message) ReadEncoding() binding.Encoding {
+	return m.encoding
+}
+
+// ReadStructured transfers a structured-mode event to a StructuredWriter.
+func (m *Message) ReadStructured(ctx context.Context, encoder binding.StructuredWriter) error {
+	if m.encoding != binding.EncodingStructured {
+		return binding.ErrNotStructured
+	}
+	return encoder.SetStructuredEvent(ctx, format.JSON, bytes.NewReader(m.Msg.Data()))
+}
+
+// ReadBinary transfers a binary-mode event to an BinaryWriter.
+func (m *Message) ReadBinary(ctx context.Context, encoder binding.BinaryWriter) error {
+	if m.encoding != binding.EncodingBinary {
+		return binding.ErrNotBinary
+	}
+
+	version := m.GetVersion()
+	if version == nil {
+		return binding.ErrNotBinary
```

Comment:
In line 66, binary encoding is already asserted by the time execution reaches line 79. It might be beneficial to use a more specific error message to clearly indicate that the version must be provided when using binary mode, such as 'Must provide the version for binary mode.' 

This will improve clarity and distinguish the error from other potential issues.

Reply:
Added a new error.